File: practicas/QytetetRuby/lib/qytetet.rb
#encoding:utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require "singleton"
require_relative "sorpresa.rb"
require_relative "tipo_sorpresa.rb"
require_relative "tablero"
require_relative "dado"
require_relative "jugador"
require_relative "casilla.rb"
require_relative "estado_juego"

module ModeloQytetet
  class Qytetet
    include Singleton
   public
    @@MAX_JUGADORES = 4
    @@NUM_CASILLAS = 20
   protected
    @@NUM_SORPRESAS = 10
    @@PRECIO_LIBERTAD = 200
    @@SALDO_SALIDA = 1000
    
    def initialize
      @cartaActual = nil
      @jugadorActual = nil
      @dado = Dado.instance
      @mazo
      @jugadores
      @tablero
      @estado = nil
    end
    #aqui no se si se deve crear esta funcion pero bueno y no se si se crea asi
    attr_reader:mazo, :jugadores, :dado, :cartaActual, :jugadorActual, :estado
  public :cartaActual, :jugadores 
    protected :mazo, :jugadorActual, :dado
  
    def setEstadoJuego(estadoJuego)
      @estado=estadoJuego
    end
    
    private
    attr_writer :cartaActual
    
    def self.MAX_JUGADORES
      return @@MAX_JUGADORES
    end
    
    def self.NUM_CASILLAS
      return @@NUM_CASILLAS
    end
    
   public
    def actuarSiEnCasillaEdificable()
      deboPagar = @jugadorActual.deboPagarAlquiler
      
      if deboPagar
        if @jugadorActual.saldo <= 0
          setEstadoJuego(EstadoJuego::ALGUNJUGADORENBANCARROTA)
        end
      end
      
      casilla = @jugadorActual.casillaActual
      tengoPropietario = casilla.tengoPropietario
      
      if @estado != EstadoJuego::ALGUNJUGADORENBANCARROTA
        
        if tengoPropietario
          setEstadoJuego(EstadoJuego::JA_PUEDEGESTIONAR)
        else
          setEstadoJuego(EstadoJuego::JA_PUEDECOMPRAROGESTIONAR)
        end
        
      end
    end
    
    def actuarSiEnCasillaNoEdificable()
      setEstadoJuego(EstadoJuego::JA_PUEDEGESTIONAR)
      
      casillaActual = @jugadorActual.casillaActual
      
      if casillaActual.tipo == TipoCasilla::IMPUESTO
        
        @jugadorActual.pagarImpuesto
        
        if casillaActual.tipo == TipoCasilla::JUEZ
          encarcelarJugadro
        else if casillaActual.tipo == TipoCasilla::SORPRESA
            cartaActual = @mazo.detete_at(0)
            setEstadoJuego(EstadoJuego::JA_CONSORPRESA)
          end
        end
      end
    end  
 
   public 
    def aplicarSorpresa()
      setEstadoJuego(EstadoJuego::JA_PUEDEGESTIONAR)
      
      if @cartaActual.tipo == TipoSorpresa::SALIRCARCEL
        @jugadorActual.setCartalibertad(@cartaActual)
      else
        if @cartaActual.tipo == TipoSorpresa::IRACASILLA
          @jugadorActual.modificarSaldo(@cartaActual.valor);
          if @jugadorActual.saldo <= 0
            setEstadoJuego(EstadoJuego::ALGUNJUGADORENBANCARROTA)
          end
        else if @cartaActual.tipo == TipoSorpresa::IRACASILLA
            valor = @cartaActual.valor
            casillaCarcel = @tablero.esCasillaCarcel(valor)
            
            if casillaCarcel == true
              encarcelarJugadro
            else
              self.mover(valor)
            end
        else if @cartaActual.tipo == TipoSorpresa::PORCASAHOTEL
            cantidad = @cartaActual.valor
            numeroTotal = @jugadorActual.cuantasCasasHoteles
            @jugadorActual.modificarSaldo(cantidad*numeroTotal)
            
            if @jugadorActual.saldo <= 0
              setEstadoJuego(EstadoJuego::ALGUNJUGADORENBANCARROTA)
            end
        else if @cartaActual.tipo == TipoSorpresa::PORJUGADOR
          @jugadores.each do |i|
            if i != @jugadorActual
              
                i.modificarSaldo(@cartaActual.valor)
              
                if i.saldo <= 0
                  setEstadoJuego(EstadoJuego::ALGUNJUGADORENBANCARROTA)
                end
                
                @jugadorActual.modificarSaldo(-@cartaActual.valor)
                
                if @jugadorActual.saldo <= 0
                  setEstadoJuego(EstadoJuego::ALGUNJUGADORENBANCARROTA)
                end
              end
            end
          end        
        end
       end
     end
    end
  end 
    
    def cancelarHipoteca(numeroCasilla)
      raise NotImplementedError
    end
    
    def comprarTituloPropiedad()
      comprado = @jugadorActual.comprarTituloPropiedad
      
      if comprado = true
        setEstadoJuego(EstadoJuego::JA_PUEDEGESTIONAR)
      end
      
      return comprado
    end
    
    def edicicarCasa(numeroCasilla)
      
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
      titulo = casilla.titulo
      edificada = @jugadorActual.edificarCasa(titulo)
      
      if edificada == true
        setEstadoJuego(EstadoJuego::JA_PUEDEGESTIONAR)
      end
      
      return edificada
    end
    
    def edificarHotel(numeroCasilla)
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
      titulo = casilla.titulo
      edificada = @jugadorActual.edificarHotel(titulo)
      
      if edificada == true
        setEstadoJuego(EstadoJuego::JA_PUEDEGESTIONAR)
      end
      
      return edificada     
    end
   private 
    def encarcelarJugadro()
      if !@jugadorActual.tengoCartaLibertad
        
        casillaCarcel = @tablero.carcel
        @jugadorActual.irACarcel(casillaCarcel)
        setEstadoJuego(EstadoJuego::JA_ENCARCELADO)
      
      else
        
        carta = @jugadorActual.devolverCartaLibertad
        @mazo<<carta
        setEstadoJuego(EstadoJuego::JA_PUEDEGESTIONAR)
        
      end
    end
    
   public 
    def getValorDado()
      return  @dado.valor
    end
    
    def hipotecarPropiedad(numeroCasilla)
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
      titulo = casilla.titulo
      @jugadorActual.hipotecarPropiedad(titulo)
      setEstadoJuego(EstadoJuego::JA_PUEDEGESTIONAR)
    end
    
   private 
    def inicializarCartasSorpresa()
      @mazo = Array.new(@@NUM_CASILLAS)
      #cobras
      @mazo<< Sorpresa.new("NOSE", 0, TipoSorpresa::PAGARCOBRAR)
      #pagas
      @mazo<< Sorpresa.new("NOSE", +10, TipoSorpresa::PAGARCOBRAR)
      #ir a casilla
      @mazo<< Sorpresa.new("NOSE", 0, TipoSorpresa::IRACASILLA)
      #ir a casilla
      @mazo<< Sorpresa.new("NOSE", 0, TipoSorpresa::IRACASILLA)
      #IR A LA CARCEL<<<<<<<<<<<<<<<<@tablero.carcel.numeroCasilla<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<ARREGLAR
      @mazo<< Sorpresa.new("NOSE", 10, TipoSorpresa::IRACASILLA)
      #recibes por casa/hotel
      @mazo<< Sorpresa.new("NOSE", 0, TipoSorpresa::PORCASAHOTEL)
      #pagas por casa/hotel
      @mazo<< Sorpresa.new("NOSE", 0, TipoSorpresa::PORCASAHOTEL)
      #recibes de jugador
      @mazo<< Sorpresa.new("NOSE", 0, TipoSorpresa::PORJUGADOR)
      #pagar a jugador
      @mazo<< Sorpresa.new("NOSE", 0, TipoSorpresa::PORJUGADOR)
      #salir carcel
      @mazo<< Sorpresa.new("NOSE", 0, TipoSorpresa::SALIRCARCEL)
    end
    
   public 
    def inicializarJuego(nombres)
      inicializarJugadores(nombres)
      inicializarTablero()
      inicializarCartasSorpresa() 
      salidaJugadores()
    end
    
   private 
    def inicializarJugadores(nombres)
      @jugadores = Array.new
       nombres.each do |i|
         @jugadores << Jugador.new(i)
       end
    end
    
    def inicializarTablero()
      @tablero=Tablero.new
    end
    
   public 
    def intentarSalirCarcel(metodo)
      if metodo == MetodoSalirCarcel::TIRANDODADO
        resultado = tirarDado()
        
        if resultado >= 5
          @jugadorActual.encarcelado(false)
        else if metodo == MetodoSalirCarcel::PAGANDOLIBERTAD
              @jugadorActual.pagarLibertad(@@PRECIO_LIBERTAD)
          end
        end
      end
      
      encarcelado = @jugadorActual.encarcelado
      
      if encarcelado 
        setEstadoJuego(EstadoJuego::JA_ENCARCELADO)
      else
        setEstadoJuego(EstadoJuego::JA_PREPARADO)
      end
      
      return !encarcelado
    end
    
    def jugar()
      desplazamiento = self.tirarDado
      numCasilla = @tablero.obtenerCasillaFinal(self.obtenerCasillaJugadorActual, desplazamiento).numeroCasilla
      self.mover(numCasilla)
    end
   
   protected 
    def mover(numCasillaDestino)
      casillaInicial = @jugadorActual.casillaActual
      casillaFinal = @tablero.obtenerCasillaNumero(numCasillaDestino)
      @jugadorActual.casillaActual(casillaFinal)
      
      if numCasillaDestino < casillaInicial.numeroCasilla
        @jugadorActual.modificarSaldo(@@SALDO_SALIDA)
      end
      
      if casillaFinal.soyEdificable
        actuarSiEnCasillaEdificable()
      else
        actuarSiEnCasillaNoEdificable()
      end
        
    end
   
   public 
    def obtenerCasillaJugadorActual()
      raise NotImplementedError
    end
    
    def obtenerCasillasTablero()
      raise NotImplementedError
    end
    
    def obtenerPropiedadesJugador()
      
      numPropiedades = Array.new
      
      @tablero.casillas.each do |i| 
         @jugadorActual.propiedades.each do |j|
            if (i.titulo == j)
                numPropiedades << i.numeroCasilla
            end
         end
      end
          
       return numPropiedades
    end
    
    def obtenerPropiedadesJugadorSegunEstadoHipoteca(estadoHipoteca)
      numPropiedades = Array.new
      
      if estadoHipoteca
        @tablero.casillas.each do |i|
          @jugadorActual.propiedades.each do |j|
            if i.itulo == j && j.hipotecada == estadoHipoteca
              numPropiedades << i.numeroCasilla
            end
          end
        end
      else
        @tablero.casillas.each do |i|
          @jugadorActual.propiedades.each do |j|
            if i.itulo == j && j.hipotecada == estadoHipoteca
              numPropiedades << i.numeroCasilla
            end
          end
        end
      end
      
      return numPropiedades
    end
    
    def obtenerRanking()
      @jugadores=@jugadores.sort
    end
    
    def obtenerSaldoJugadorActual()
      return @jugadorActual.saldo
    end
   
   private 
    def salidaJugadores()
      for i in 0..@jugadores.size
        @jugadores[i].casillaActual(@tablero.btenerCasillaNumero(0))
      end
      
      r = Random.new
      @jugadorActual = @jugadoes[r.rand(@jugadores.size)]
      
      @estado = EstadoJuego::JA_PREPARADO
    end
   
   public 
    def siguienteJugador()
      siguente = @jugadores.index(@jugadorActual)
      siguiente = (siguiente + 1) % @jugadores.size
      @jugadorActual = @jugadores[siguiente]
      
      if @jugadorActual.encarcelado
        @estado = EstadoJuego::JA_ENCARCELADOCONOPCIONDELIBERTAD
      else
        @estado = EstadoJuego::JA_PREPARADO
      end
    end
   
   protected 
    def tirarDado()
      return @dado.tirar
    end
   
   public 
    def venderPropiedad(numeroCasilla)
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
      @jugadorActual.venderPropiedad(casilla)
      setEstadoJuego(EstadoJuego::JA_PUEDEGESTIONAR)
    end
  
    def to_s
      "Qytetet: \n carta actual: #{@cartaActual}  \n jugador actual: #{@jugadorActual} \n dado: #{@dado} \n mazo:{ #{@mazo.to_s} } \n jugadores: \n { #{@jugadores} } \n tablero: \n { #{@tablero} }}"
    end
  end
end

File: QytetetJava/src/modeloqytetet/Qytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Random;
/**
 *
 * @author venrra
 */
public class Qytetet {
    public static int MAX_JUGADORES = 4;
    static int NUM_SORPRESAS = 10;
    public static int NUM_CASILLAS = 20;
    static int PRECIO_LIBERTAD = 200;
    static int SALDO_SALIDA = 1000;
    
    private ArrayList<Sorpresa> mazo;
    private Tablero tablero;
    private Sorpresa cartaActual;
    private ArrayList<Jugador> jugadores;
    private Jugador jugadorActual;
    private Dado dado;
    private EstadoJuego estado;
    
    private Qytetet() {
        this.cartaActual = null;
        this.jugadorActual = null;
        this.dado = Dado.getinstance();
        //this.estado=
    }
    
    private static final Qytetet instance = new Qytetet();
    
    public static Qytetet getInstance(){
        return instance;
    }
    
    void actuarSiEnCasillaEdificable(){
        boolean deboPagar = jugadorActual.deboPagarAlquiler();
        
        if (deboPagar){
            jugadorActual.pagarAlquiler();
            if(jugadorActual.getSaldo()<=0)
                this.setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
        }
        
        Casilla casilla = jugadorActual.getCasillaActual();
        boolean tengoPropietario = casilla.tengoPropietario();
        
        if (this.estado != EstadoJuego.ALGUNJUGADORENBANCARROTA){
            
            if(tengoPropietario)
                this.setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
            else
                this.setEstadoJuego(EstadoJuego.JA_PUEDECOMPRAROGESTIONAR);
        }
    }
    
    void actuarSiEnCasillaNoEdificable(){
        this.setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        
        Casilla casillaActual = this.jugadorActual.getCasillaActual();
        
        if (casillaActual.getTipo() == TipoCasilla.IMPUESTO){
            
            this.jugadorActual.pagarImpuesto();
            
            if (casillaActual.getTipo() == TipoCasilla.JUEZ){
                
                this.encarcelarJugador();
            
            }else if(casillaActual.getTipo() == TipoCasilla.SORPRESA){
                cartaActual = mazo.remove(0);
                this.setEstadoJuego(EstadoJuego.JA_CONSORPRESA);
            }
        }
        
    }
    
    public void aplicarSorpresa(){
        this.setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        
        if (this.cartaActual.getTipo() == TipoSorpresa.SALIRCARCEL){
            
            this.jugadorActual.setCartalibertad(cartaActual);
        
        }else{
            if(this.cartaActual.getTipo() == TipoSorpresa.IRACASILLA){
                
                this.jugadorActual.modificarSaldo(cartaActual.getValor());
                
                if(this.jugadorActual.getSaldo() <= 0)
                    this.setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
                    
            }else if(this.cartaActual.getTipo() == TipoSorpresa.IRACASILLA){
                    
                int valor = cartaActual.getValor();
                boolean casillaCarcel = tablero.esCasillaCarcel(valor);
                
                if(casillaCarcel == true)
                    this.encarcelarJugador();
                else
                    this.mover(valor);
                
            }else if(this.cartaActual.getTipo() == TipoSorpresa.PORCASAHOTEL){
                
               int cantidad = this.cartaActual.getValor();
               
               int numeroTotal = this.jugadorActual.cuantasCasasHoteles();
               
               this.jugadorActual.modificarSaldo(cantidad*numeroTotal);
               
               if(this.jugadorActual.getSaldo() <= 0)
                   this.setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
               
            }else if(this.cartaActual.getTipo() == TipoSorpresa.PORJUGADOR){
                for(Jugador i : jugadores){
                    
                    if(i != jugadorActual){
                        
                        i.modificarSaldo(cartaActual.getValor());
                        
                        if (i.getSaldo() <= 0)
                            this.setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
                        
                        this.jugadorActual.modificarSaldo(-this.cartaActual.getValor());
                        
                        if (this.jugadorActual.getSaldo() <= 0)
                            this.setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
                    }
                }
            }
        }
    }
    
    public boolean cancelarHipoteca(int numeroCasilla){
        throw new UnsupportedOperationException("Sin implementar");
    }
    
    public boolean comprarTituloPropiedad(){
        boolean comprado = this.jugadorActual.comprarTituloPropiedad();
        
        if(comprado == true)
            this.setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        
        return comprado;
        
    }
    
    public boolean edificarCasa(int numeroCasilla){
        
        Casilla casilla = tablero.obtenerCasillaNumero(numeroCasilla);
        TituloPropiedad titulo = casilla.getTitulo();
        boolean edificada = this.jugadorActual.edificarCasa(titulo);
        
        if(edificada == true)
            this.setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        
        return edificada;
    }
    
    public boolean edificarHotel(int numeroCasilla){
        Casilla casilla = tablero.obtenerCasillaNumero(numeroCasilla);
        TituloPropiedad titulo = casilla.getTitulo();
        boolean edificada = this.jugadorActual.edificarHotel(titulo);
        
        if (edificada == true)
            this.setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        
        return edificada;
    }
    
    private void encarcelarJugador(){
        if (!this.jugadorActual.tengoCartaLibertad()){
           
            Casilla casillaCarcel = tablero.getCarcel();
            this.jugadorActual.irACarcel(casillaCarcel);
            this.setEstadoJuego(EstadoJuego.JA_ENCARCELADO);
            
        }else{
            
            Sorpresa carta = this.jugadorActual.devolverCartaLibertad();
            mazo.add(carta);
            this.setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);

        }
    }

    public Sorpresa getCartaActual() {
        return cartaActual;
    }
    
    Dado getDado() {
        return dado;
    }    

    Jugador getJugadorActual() {
        return jugadorActual;
    }
    
    public ArrayList<Jugador> getJugadores() {
        return jugadores;
    }

    public Tablero getTablero() {
        return tablero;
    }
     
    ArrayList<Sorpresa> getMazo(){
        return this.mazo;
    }
    
    public int getValorDado(){
        return this.dado.getValor();
    }

    public void hipotecarPropiedades(int numeroCasilla){
        Casilla casilla = tablero.obtenerCasillaNumero(numeroCasilla);
        TituloPropiedad titulo = casilla.getTitulo();
        this.jugadorActual.hipotecarPropiedad(titulo);
        this.setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
    }
    
    private void inicializarCartasSorpresa(){
        mazo = new ArrayList<>();
        //cobras 
        mazo.add(new Sorpresa ("NOSE", 200, TipoSorpresa.PAGARCOBRAR));
        //pagas
        mazo.add(new Sorpresa ("NOSE", -1000, TipoSorpresa.PAGARCOBRAR));
        //ir a casilla
        mazo.add(new Sorpresa ("CASILLA1", 1, TipoSorpresa.IRACASILLA));
        //ir a casilla
        mazo.add(new Sorpresa ("CASILLA19", 19, TipoSorpresa.IRACASILLA));
        //ir a la carcel
        mazo.add(new Sorpresa ("CARCEL", tablero.getCarcel().getNumeroCasilla(), TipoSorpresa.IRACASILLA));
        //cobra por las casa y hoteles que tenga
        mazo.add(new Sorpresa ("NOSE", 150, TipoSorpresa.PORCASAHOTEL));
        //paga por las casa y hoteles que tenga
        mazo.add(new Sorpresa ("NOSE", -100, TipoSorpresa.PORCASAHOTEL));
        //recibes 
        mazo.add(new Sorpresa ("NOSE", 50, TipoSorpresa.PORJUGADOR));
        //pagas 
        mazo.add(new Sorpresa ("NOSE", 75, TipoSorpresa.PORJUGADOR));
        //SALR DE LA CARCEL
        mazo.add(new Sorpresa ("NOSE", 0,TipoSorpresa.SALIRCARCEL));
    }
       
    public void inicializarJuego(ArrayList<String> nombres){
        inicializarJugadores(nombres);
        inicializarTablero();
        inicializarCartasSorpresa();
        salidaJugadores();
    }
    
    private void inicializarJugadores(ArrayList<String> nombres){
        jugadores = new ArrayList<>();
        for(String t: nombres)
            this.jugadores.add(new Jugador(t));
    }
    
    private void inicializarTablero(){
        this.tablero = new Tablero();
    }

    public boolean intentarSalirCarcel(MetodoSalirCarcel metodo){
        if(metodo == MetodoSalirCarcel.TIRANDODADO){
            int resultado = this.tirarDado();
           
            if(resultado >= 5){
                
                this.jugadorActual.setEncarcelado(false);
           
            }else if(metodo == MetodoSalirCarcel.PAGANDOLIBERTAD){
                
                this.jugadorActual.pagarLibertad(PRECIO_LIBERTAD);
                
            }
        }
            
        boolean encarcelado = this.jugadorActual.getEncarcelado();

        if(encarcelado)
            this.setEstadoJuego(EstadoJuego.JA_ENCARCELADO);
        else
            this.setEstadoJuego(EstadoJuego.JA_PREPARADO);
        
        return !encarcelado;
    }
            
    public void jugar(){
        int desplazamiento = this.tirarDado();
        int numCasilla = this.tablero.obtenerCasillaFinal(this.obtenerCasillaJugadorActual(), desplazamiento).getNumeroCasilla();
        this.mover(numCasilla);
    }
    
    void mover(int numCasillaDestino){
        
        Casilla casillaInicial = jugadorActual.getCasillaActual();
        Casilla casillaFinal = tablero.obtenerCasillaNumero(numCasillaDestino);
        this.jugadorActual.setCasillaActual(casillaFinal);
        
        if(numCasillaDestino < casillaInicial.getNumeroCasilla())
            this.jugadorActual.modificarSaldo(SALDO_SALIDA);
        
        if(casillaFinal.soyEdificable())
            this.actuarSiEnCasillaEdificable();
        else
            this.actuarSiEnCasillaNoEdificable();

    }

    public Casilla obtenerCasillaJugadorActual(){
        throw new UnsupportedOperationException("Sin implementar");
    }
    
    public ArrayList<Casilla> obtenerCasillasTablero(){
        throw new UnsupportedOperationException("Sin implementar");
    }
    
    public ArrayList<Integer> obtenerPropiedadesJugador(){
        ArrayList<Integer> numPropiedades = new ArrayList<>();
        for (Casilla i: tablero.getCasillas())
            for(TituloPropiedad j : jugadorActual.getPropiedades())
                if (i.getTitulo() == j)
                    numPropiedades.add(i.getNumeroCasilla());
        
        return numPropiedades;
    }
    
    public ArrayList<Integer> obtenerPropidadesJugadorSegunEstadoHipoteca(boolean estadoHipoteca){
        ArrayList<Integer> numPropiedades = new ArrayList<>();        
        if (estadoHipoteca){
            for (Casilla i: tablero.getCasillas())
                for(TituloPropiedad j : jugadorActual.getPropiedades())
                    if (i.getTitulo() == j && j.getHipotecada() == estadoHipoteca)
                        numPropiedades.add(i.getNumeroCasilla());
        }else
            for (Casilla i: tablero.getCasillas())
                for(TituloPropiedad j : jugadorActual.getPropiedades())
                    if (i.getTitulo() == j && j.getHipotecada() == estadoHipoteca)
                        numPropiedades.add(i.getNumeroCasilla());
        
        return numPropiedades;
    }
    
    public void obtenerRanking(){
        Collections.sort(this.jugadores);
    }
    
    public int obtenerSaldoJugadorActual(){
        return jugadorActual.getSaldo();
    }
    
    private void salidaJugadores(){
        for (int i=0; i < jugadores.size(); i++)
            jugadores.get(i).setCasillaActual(tablero.obtenerCasillaNumero(0));
        
        Random r = new Random();
        jugadorActual = jugadores.get( r.nextInt( jugadores.size() ));
        
        estado = EstadoJuego.JA_PREPARADO;
    }

    public void setCartaActual(Sorpresa cartaActual) {
        this.cartaActual = cartaActual;
    }
    
    void setEstadoJuego(EstadoJuego estadoJuego){
        this.estado=estadoJuego;
    }
    
    public void siguienteJugador(){
        int siguiente = jugadores.indexOf(jugadorActual);
        siguiente = (siguiente + 1) % jugadores.size();
        jugadorActual = jugadores.get(siguiente);
        
        if (jugadorActual.getEncarcelado())
            estado = EstadoJuego.JA_ENCARCELADOCONOPCIONDELIBERTAD;
        else 
            estado = EstadoJuego.JA_PREPARADO;
    }
    
    int tirarDado(){
        return this.dado.tirar();
    }
    
    public void venderPropiedad(int numeroCasilla){
        Casilla casilla = tablero.obtenerCasillaNumero(numeroCasilla);
        
        this.jugadorActual.venderPropiedad(casilla);
        
        this.setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
    }

    @Override
    public String toString() {
        return "Qytetet{" + "mazo=" + mazo + ", tablero=" + tablero + ", cartaActual=" + cartaActual + ",\n jugadores=\n" + jugadores + ", jugadorActual=" + jugadorActual + ", dado=" + dado + '}';
    }
    
    
}